crates/ty_python_semantic/src/types.rs

```diff
@@ -4486,9 +4486,10 @@ impl<'db> Type<'db> {
         //    fallback to `object.__init__`, since it will indeed check that no arguments are
         //    passed.
         //
-        // Note that we currently ignore `__new__` return type, since we do not yet support `Self`
-        // and most builtin classes use it as return type annotation. We always return the instance
-        // type.
+        // Note that we currently ignore `__new__` return type in most cases, since we do not yet
+        // support `Self` and most builtin classes use it as return type annotation.
```

Comment:
We do support `Self` now. (We don't yet assume it as the type of an un-annotated `self` arg, but for `__new__` that's not relevant anyway.) So this PR should explicitly specialize a `__new__` return type of `Self` to the `instance_ty`, and update this comment to remove the mention of not supporting `Self`.